Qt: display game icon and title in the game list

wwylele opened the pull request:
![list](https://cloud.githubusercontent.com/assets/4592895/14751884/fc2ad160-08d4-11e6-838b-52a61ecd69ea.PNG)

@JamePeng told me this makes it much easier to find games.

mailwl commented:
Nice, but "File type" column looks ugly now. Maybe replace it with images too, or shift between name and size?

lioncash commented on src/common/common_funcs.h:
```diff
@@ -87,3 +87,7 @@ inline u64 _rotr64(u64 x, unsigned int shift){
 // Defined in Misc.cpp.
 const char* GetLastErrorMsg();
+
+constexpr u32 MakeMagic(char a, char b, char c, char d) {
```
This should likely be put elsewhere preferably in a place with a namespace. Ideally common_funcs shouldn't even exist as a header, since it pulls in a bunch of unrelated stuff when its included.

lioncash commented on src/citra_qt/game_list_p.h:
```diff
@@ -27,29 +63,50 @@ class GameListItem : public QStandardItem {
+        if (MakeMagic('S', 'M', 'D', 'H') != smdh->magic) return;
+
+        // Get icon from smdh
+        QImage icon(48, 48, QImage::Format::Format_RGB888);
```
Image retrieval should be its own function

lioncash commented on src/citra_qt/game_list_p.h:
```diff
@@ -27,29 +63,50 @@ class GameListItem : public QStandardItem {
+        setData(QPixmap::fromImage(icon), Qt::DecorationRole);
+
+        // Get title form smdh
+        setData(QString::fromUtf16(smdh->titles[1].short_title.data()), TitleRole);
```
Title retrieval should be its own function.

lioncash commented on src/citra_qt/game_list_p.h:
```diff
@@ -27,29 +63,50 @@ class GameListItem : public QStandardItem {
     {
         setData(game_path, FullPathRole);
+
+        if (smdh_data.size() < sizeof(SMDH)) return;
```
Sanitizing should also be its own function (`IsValidSMDH or something')

lioncash commented on src/citra_qt/game_list.cpp:
```diff
@@ -143,9 +145,17 @@ void GameListWorker::AddFstEntriesToGameList(const std::string& dir_path, bool d
 
+            std::vector<u8> smdh;
+            std::unique_ptr<Loader::AppLoader> loader;
+            if (filetype == Loader::FileType::CCI || filetype==Loader::FileType::CXI) {
```
Acquiring the loader should be its own function as well.

lioncash commented on src/citra_qt/game_list.cpp:
```diff
@@ -11,6 +11,8 @@
 
 #include "core/loader/loader.h"
+#include "core/loader/ncch.h"
+#include "core/loader/3dsx.h"
```
3dsx.h comes before loader.h

lioncash commented on src/citra_qt/game_list_p.h:
```diff
@@ -27,29 +63,50 @@ class GameListItem : public QStandardItem {
         setData(game_path, FullPathRole);
+
+        if (smdh_data.size() < sizeof(SMDH)) return;
+        const SMDH* smdh = reinterpret_cast<const SMDH*>(smdh_data.data());
```
This reinterpret cast is undefined behavior.

lioncash commented on src/core/loader/3dsx.cpp:
```diff
@@ -303,4 +303,29 @@ ResultStatus AppLoader_THREEDSX::ReadRomFS(std::shared_ptr<FileUtil::IOFile>& ro
     return ResultStatus::ErrorNotUsed;
 }
 
+ResultStatus AppLoader_THREEDSX::ReadIcon(std::vector<u8>& buffer) {
```
You can use boost::optional here to avoid using an out parameter.

wwylele replied:
It is a virtual function in `AppLoader` (though never used before). Should I still change it?

bunnei replied:
I think it's fine, refactoring the interface (which is used elsewhere in AppLoader) is out of scope of this PR, IMO

lioncash commented on src/core/loader/ncch.cpp:
```diff
@@ -173,6 +173,11 @@ ResultStatus AppLoader_NCCH::LoadSectionExeFS(const char* name, std::vector<u8>&
     if (!file.IsOpen())
         return ResultStatus::Error;
 
+    if (!is_exefs_loaded) {
```
This should likely be its own function of some sort, it's copied twice.

wwylele commented:
![default](https://cloud.githubusercontent.com/assets/4592895/14760759/e14eb4c6-0954-11e6-87b6-92e0b0637f2a.PNG)
Update: swap name and type column, and add a blank icon to those without SMDH data

LittleWhite-tb commented:
Really nice feature :+1:

lioncash commented on src/citra_qt/game_list_p.h:
```diff
@@ -5,13 +5,128 @@
+
+    ///an enum indicating the language for each title entry
+    enum class TitleLanguage {
+        JAPANESE = 0,
```
These should be the regularly cased. e.g. `Japanese`, `English`, etc.

lioncash commented on src/citra_qt/game_list_p.h:
```diff
@@ -5,13 +5,128 @@
+        return icon;
+    }
+
+    ///an enum indicating the language for each title entry
```
`/// Indicates the language used for each title entry`

The fact it's an enum is self-evident.

lioncash commented on src/citra_qt/game_list_p.h:
```diff
@@ -5,13 +5,128 @@
+     * @param language title language
+     * @return QString short title
+     */
+    QString GetShortTitle(TitleLanguage language) {
```
`Gets the short game title from the SMDH`

Also this function can be `const`

lioncash commented on src/citra_qt/game_list_p.h:
```diff
@@ -5,13 +5,128 @@
+     */
+    static QPixmap GetDefaultIcon(bool large) {
+        int size = large ? 48 : 24;
+        QPixmap icon(size,size);
```
Space after the comma

lioncash commented on src/citra_qt/game_list_p.h:
```diff
@@ -5,13 +5,128 @@
+    }
+
+    /**
+     * Get default icon (for games without valid SMDH)
```
`Gets the default icon (for games without a valid SMDH)`

lioncash commented on src/citra_qt/game_list_p.h:
```diff
@@ -5,13 +5,128 @@
+        int size = large ? 48 : 24;
+        QPixmap icon(size,size);
+        //Fill with white
+        icon.fill();
```
Unless our background is explicitly set to white, this'll show a white block on OSes (or OS themes) that have a different background scheme

lioncash commented on src/citra_qt/game_list_p.h:
```diff
@@ -5,13 +5,128 @@
+     * @return bool test result
+     */
+    static bool IsValidSMDH(const std::vector<u8>& smdh_data) {
+        if (smdh_data.size() < sizeof(SMDH)) return false;
```
```cpp
if (smdh_data.size() < sizeof(SMDH))
    return false;
```

lioncash commented on src/citra_qt/game_list_p.h:
```diff
@@ -5,13 +5,128 @@
 #include "common/string_util.h"
+#include "common/common_funcs.h"
+#include "common/common_types.h"
+#include "common/color.h"
```
color comes before common

lioncash commented on src/citra_qt/game_list_p.h:
```diff
@@ -5,13 +5,128 @@
 #include <QString>
+#include <QImage>
+
+#include "core/loader/loader.h"
```
core comes after common includes

lioncash commented on src/citra_qt/game_list_p.h:
```diff
@@ -5,13 +5,128 @@
 #include <QRunnable>
 #include <QStandardItem>
 #include <QString>
+#include <QImage>
```
This should be above QRunnable.

lioncash commented on src/citra_qt/game_list_p.h:
```diff
@@ -27,29 +142,46 @@ class GameListItem : public QStandardItem {
+        SMDH smdh;
+        memcpy(&smdh, smdh_data.data(), sizeof(SMDH));
+
+        // Get icon from smdh
```
SMDH

lioncash commented on src/citra_qt/game_list_p.h:
```diff
@@ -27,29 +142,46 @@ class GameListItem : public QStandardItem {
+        // Get icon from smdh
+        setData(smdh.GetIcon(true), Qt::DecorationRole);
+
+        // Get title form smdh
```
SMDH

lioncash commented on src/core/loader/loader.cpp:
```diff
@@ -90,6 +90,28 @@ const char* GetFileTypeString(FileType type) {
+    case FileType::ELF:
+        return std::make_unique<AppLoader_ELF>(std::move(file), filename);
+
+    // NCCH/NCSD container formats...
```
the `...` are unnecessary

lioncash commented on src/core/loader/loader.h:
```diff
@@ -150,6 +150,17 @@ class AppLoader : NonCopyable {
+ * Get a loader for a file with a specific type
+ * @param file The file to load
+ * @param type The type of the file
+ * @filename the file name (without path)
```
`@param`

lioncash commented on src/core/loader/loader.h:
```diff
@@ -150,6 +150,17 @@ class AppLoader : NonCopyable {
+ * @param file The file to load
+ * @param type The type of the file
+ * @filename the file name (without path)
+ * @filepath the file full path (with name)
```
@param

lioncash commented on src/core/loader/loader.h:
```diff
@@ -150,6 +150,17 @@ class AppLoader : NonCopyable {
 extern const std::initializer_list<Kernel::AddressMapping> default_address_mappings;
 
 /**
+ * Get a loader for a file with a specific type
```
There should be a blank newline after this to separate the description from the parameter documentation

wwylele replied:
@lioncash all other documentation around don't have blank line separating the description from the parameter... should I still add it? and to all other documentation?

lioncash replied:
Oh in that case don't worry about it, then.

lioncash commented on src/core/loader/ncch.cpp:
```diff
@@ -173,6 +173,9 @@ ResultStatus AppLoader_NCCH::LoadSectionExeFS(const char* name, std::vector<u8>&
         return ResultStatus::Error;
 
+    ResultStatus r = LoadExeFS();
+    if (r != ResultStatus::Success) return r;
```
```cpp
if (r != ResultStatus::Success)
    return r;
```

lioncash commented on src/core/loader/ncch.cpp:
```diff
@@ -282,6 +285,17 @@ ResultStatus AppLoader_NCCH::Load() {
+        return ResultStatus::ErrorAlreadyLoaded;
+
+    ResultStatus r = LoadExeFS();
+    if (r != ResultStatus::Success) return r;
```
```cpp
if (r != ResultStatus::Success)
    return r;
```

wwylele commented:
@lioncash Thank you very much for reviewing! By the way, I think the [Contributing guide](https://github.com/citra-emu/citra/blob/master/CONTRIBUTING.md) needs update. Something there is unclear ( e.g. the naming rule for enum).

lioncash commented on src/citra_qt/game_list_p.h:
```diff
@@ -5,13 +5,128 @@
 #pragma once
 
 #include <atomic>
+#include <array>
```
Sorry, I missed this one, this should be above `<atomic>`

mailwl commented:
@wwylele, may i ask you as Qt-guru create toolbar with all this thigger buttons, which now in "Configure.." menu button? I hate this redundant clicks to get needed option with new Citra Configuration window.

Sorry for offtop

wwylele commented:
@mailwl technically it's not hard, but I don't know if everyone like that... if you really want it, I can make it for you on my branch someday.

mailwl commented:
hmm... maybe add to menu add option to show/hide toolbar? I think, this will be useful even in future release state of Citra.
eh, it's not the place for discussion offtopics

JayFoxRox commented on src/citra_qt/game_list.cpp:
```diff
@@ -34,8 +34,8 @@ GameList::GameList(QWidget* parent)
 
     item_model->insertColumns(0, COLUMN_COUNT);
-    item_model->setHeaderData(COLUMN_FILE_TYPE, Qt::Horizontal, "File type");
     item_model->setHeaderData(COLUMN_NAME, Qt::Horizontal, "Name");
```
Can you tell this column to use the available space or something? The default settings for me were that it only had enough room for the icon, I couldn't even see the games name.

wwylele replied:
@JayFoxRox added, but not sure if it's in a correct way.

bunnei commented on src/citra_qt/game_list.cpp:
```diff
@@ -143,9 +147,12 @@ void GameListWorker::AddFstEntriesToGameList(const std::string& dir_path, bool d
 
+            std::vector<u8> smdh;
+            std::unique_ptr<Loader::AppLoader> loader = Loader::GetLoader(FileUtil::IOFile(physical_name, "rb"), filetype, filename_filename, physical_name);
+            if (loader) loader->ReadIcon(smdh);
```
new line

bunnei commented on src/citra_qt/game_list.cpp:
```diff
@@ -143,9 +147,12 @@ void GameListWorker::AddFstEntriesToGameList(const std::string& dir_path, bool d
+            std::vector<u8> smdh;
+            std::unique_ptr<Loader::AppLoader> loader = Loader::GetLoader(FileUtil::IOFile(physical_name, "rb"), filetype, filename_filename, physical_name);
+            if (loader) loader->ReadIcon(smdh);
             emit EntryReady({
```
Could use a space before this and the statements above

bunnei commented on src/core/loader/ncch.h:
```diff
@@ -232,6 +232,14 @@ class AppLoader_NCCH final : public AppLoader {
+    ResultStatus LoadExeFS();
+
+    bool            is_exefs_loaded = false;
+
```
Unnecessary extra space

bunnei commented on src/core/loader/ncch.cpp:
```diff
@@ -282,6 +286,18 @@ ResultStatus AppLoader_NCCH::Load() {
+    if (is_loaded)
+        return ResultStatus::ErrorAlreadyLoaded;
+
+    ResultStatus r = LoadExeFS();
```
Use `result` instead of `r`. Single letter variable names are easily abused and are against our style guide, unless for a loop index e.g. i

bunnei commented on src/core/loader/ncch.cpp:
```diff
@@ -173,6 +173,10 @@ ResultStatus AppLoader_NCCH::LoadSectionExeFS(const char* name, std::vector<u8>&
     if (!file.IsOpen())
         return ResultStatus::Error;
 
+    ResultStatus r = LoadExeFS();
```
Ditto on single letter variable name

bunnei commented on src/core/loader/loader.h:
```diff
@@ -150,6 +150,17 @@ class AppLoader : NonCopyable {
+ * @return std::unique_ptr<AppLoader> a pointer to a loader object;  null pointer for unsupported type
+ */
+std::unique_ptr<AppLoader> GetLoader(FileUtil::IOFile&& file, FileType type,
+    const std::string& filename, const std::string& filepath);
```
Unnecessary line break here

bunnei commented on src/core/loader/loader.h:
```diff
@@ -150,6 +150,17 @@ class AppLoader : NonCopyable {
+ * @param type The type of the file
+ * @param filename the file name (without path)
+ * @param filepath the file full path (with name)
+ * @return std::unique_ptr<AppLoader> a pointer to a loader object;  null pointer for unsupported type
```
*nullptr

bunnei commented on src/core/loader/3dsx.cpp:
```diff
@@ -303,4 +303,29 @@ ResultStatus AppLoader_THREEDSX::ReadRomFS(std::shared_ptr<FileUtil::IOFile>& ro
+    if (hdr.smdh_offset != 0) {
+        file.Seek(hdr.smdh_offset, SEEK_SET);
+        buffer.resize(hdr.smdh_size);
+        if (file.ReadBytes(&buffer[0], hdr.smdh_size) != hdr.smdh_size)
```
Space before this line

bunnei commented on src/core/loader/3dsx.cpp:
```diff
@@ -303,4 +303,29 @@ ResultStatus AppLoader_THREEDSX::ReadRomFS(std::shared_ptr<FileUtil::IOFile>& ro
+        file.Seek(hdr.smdh_offset, SEEK_SET);
+        buffer.resize(hdr.smdh_size);
+        if (file.ReadBytes(&buffer[0], hdr.smdh_size) != hdr.smdh_size)
+            return ResultStatus::Error;
```
Space after this line. Don't make your code so crammed!

bunnei commented on src/citra_qt/game_list_p.h:
```diff
@@ -27,29 +142,46 @@ class GameListItem : public QStandardItem {
-            GameListItem::setData(value, FullPathRole);
+            Common::SplitPath(data(FullPathRole).toString().toStdString(), nullptr, &filename, nullptr);
+            QString title = data(TitleRole).toString();
+            if (title.isEmpty())
```
Use a ternary.. e.g. `return title.isEmpty() ? foo : bar;`

bunnei commented on src/citra_qt/game_list_p.h:
```diff
@@ -4,14 +4,129 @@
+
+    /**
+     * Gets game icon from SMDH
+     * @param large set true to get large icon (48x48), false to get small icon (24x24)
```
Grammar nit:
"If true, returns large icon (48x48), otherwise returns small icon (24x24)"

bunnei commented on src/citra_qt/game_list_p.h:
```diff
@@ -4,14 +4,129 @@
+    std::array<u8, 0x1200> large_icon;
+
+    /**
+     * Tests if data is a valid SMDH by its length and magic.
```
"magic number"

bunnei commented on src/citra_qt/game_list_p.h:
```diff
@@ -4,14 +4,129 @@
+     */
+    static bool IsValidSMDH(const std::vector<u8>& smdh_data) {
+        if (smdh_data.size() < sizeof(SMDH))
+            return false;
```
Space after this line

bunnei commented on src/citra_qt/game_list_p.h:
```diff
@@ -4,14 +4,129 @@
+        if (smdh_data.size() < sizeof(SMDH))
+            return false;
+        u32 magic;
+        memcpy(&magic, smdh_data.data(), 4);
```
Space after this line

bunnei commented on src/citra_qt/game_list_p.h:
```diff
@@ -4,14 +4,129 @@
+#include "video_core/utils.h"
+
+/// SMDH data structure that contains title, titles etc. See https://www.3dbrew.org/wiki/SMDH
+struct SMDH {
```
This code to decode SMDH icons does not belong in the Qt frontend, it should be in core since it's part of 3DS emulation. Probably belongs in core/loader/ncch.*, since this icon format is specific to CCI/CXI.

bunnei commented:
Code generally looks fine, aside from the nits and the restructure to move icon decoding to core/loader. Some other general comments:
* I don't see the need to print filenames in the list view, I would just print the game name. We probably will (at some point) want to have an internal DB of nice names, but for now I think just doing this is much nicer than file names.
* I already noticed I have some duplicate filenames in my library for games that I have multiple regions of. Not critical for this PR, but if you're interested in continuing to work on the UI, it would be awesome to also have a region column in the game list view.

Anyway, this looks awesome and I hope to get it merged soon!

wwylele commented:
Addressed @bunnei 's comment. And:
- I moved SMDH structure to Loader.h because 3dsx also contains data in that format. Also for reducing header including in game_list_p.h
- After this PR merged I will work on adding region info. (we are really getting identical to dolphin's game list :sweat_smile: )
- I want to keep file names, at least for now, for the following reason:
 -  It is useful as an extra description for each game file. For instance, I give some games file name with region and game ID, so that I can distinguish same game from different region, and get ID easily when I need it.
 - The name column of game list is currently functioning not only for displaying, but also for storing file path for booting. It's a little bit weird to store file name but not to display them. So it needs some reconstruction if it doesn't display file name. I want to leave the for next PR (maybe along with region display)

wwylele commented:
Found game_list_p.h was missing in CMakeList, so I added it. Sorry for strange commit order - my computer's system clock just screwed up.

MerryMage commented on src/citra_qt/game_list.cpp:
```diff
@@ -109,7 +109,11 @@ void GameList::SaveInterfaceLayout()
+    if (!header->restoreState(UISettings::values.gamelist_header_state)) {
+        // We are using the name column to display icons and titles
+        // so make it as large as possible as default.
+        header->resizeSection(COLUMN_NAME, header->width());
```
You should likely use [`setSectionResizeMode`](http://doc.qt.io/qt-5/qheaderview.html#setSectionResizeMode) with `QHeaderView::Stretch`.

wwylele replied:
@MerryMage If I set it to `Stretch`, user won't be able to resize the column.

bunnei commented on src/citra_qt/game_list_p.h:
```diff
@@ -27,29 +99,43 @@ class GameListItem : public QStandardItem {
  * A specialization of GameListItem for path values.
  * This class ensures that for every full path value it holds, a correct string representation
  * of just the filename (with no extension) will be displayed to the user.
+ * If this class recieves a valid SMDH data, it will also display game icons and titles.
```
*receives valid (no 'a')

bunnei commented:
Code looks fine to me.

Last call for additional reviews on this one.

Subv commented:
LGTM

bunnei commented:
LGTM as well, ofc
